[cutlass backend] Check if len(timings) == len(choices) before skipping precompile

henrylhtsang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #148050

Differential Revision: [D70298908](https://our.internmc.facebook.com/intern/diff/D70298908/)

Mostly from @coconutruben observation. Right now, we skip precompilation if we find **some** timings. That sounds like a bug. Most of the time it is fine, since we don't change the number of configs and triton compilation doesn't take too long. But it is devastating for cutlass backend.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

henrylhtsang committed:
[cutlass backend] Check if len(timings) == len(choices) before skipping precompile



Differential Revision: [D70298908](https://our.internmc.facebook.com/intern/diff/D70298908/)

[ghstack-poisoned]

henrylhtsang committed:
Update on "[cutlass backend] Check if len(timings) == len(choices) before skipping precompile"


Differential Revision: [D70298908](https://our.internmc.facebook.com/intern/diff/D70298908/)

Mostly from coconutruben observation. Right now, we skip precompilation if we find **some** timings. That sounds like a bug. Most of the time it is fine, since we don't change the number of configs and triton compilation doesn't take too long. But it is devastating for cutlass backend.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

coconutruben commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1757,8 +1757,16 @@ def no_op(*args, **kwargs):
+                if len(timings) == len(choices):
+                    log.debug("Timings found in cache, returning no_op")
+                    return no_op
+                else:
```
nit: add a comment here and confirm that I think even if you "precompile" the same choice twice, it's actually very fast, because CodeCache will see the hash and skip anyways. This is to help future readers understand that you don't need something fancy here to only precompile the choices that are missing, but that this simple check is sufficient

henrylhtsang replied:
> nit: add a comment here and confirm that I think even if you "precompile" the same choice twice, it's actually very fast, because CodeCache will see the hash and skip anyways. This is to help future readers understand that you don't need something fancy here to only precompile the choices that are missing, but that this simple check is sufficient

Added a side comment, but I think I will leave out the codecache part, since I think its awkward to explain caching logic there

henrylhtsang committed:
Update on "[cutlass backend] Check if len(timings) == len(choices) before skipping precompile"


Differential Revision: [D70298908](https://our.internmc.facebook.com/intern/diff/D70298908/)

Mostly from coconutruben observation. Right now, we skip precompilation if we find **some** timings. That sounds like a bug. Most of the time it is fine, since we don't change the number of configs and triton compilation doesn't take too long. But it is devastating for cutlass backend.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -1757,8 +1757,20 @@ def no_op(*args, **kwargs):
+                else:
+                    # TODO: remove this branch in April 2025
+                    # added warning for debugging purpose
+                    log.warning(
```
We should avoid adding warnings that users might see which are intended for developers. I think warning will be visible even without torch logs set. can you use info instead ?

henrylhtsang committed:
Update on "[cutlass backend] Check if len(timings) == len(choices) before skipping precompile"


Differential Revision: [D70298908](https://our.internmc.facebook.com/intern/diff/D70298908/)

Mostly from coconutruben observation. Right now, we skip precompilation if we find **some** timings. That sounds like a bug. Most of the time it is fine, since we don't change the number of configs and triton compilation doesn't take too long. But it is devastating for cutlass backend.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

henrylhtsang commented:
@pytorchbot merge -i

henrylhtsang commented:
@pytorchbot merge -i
